### PR TITLE
Ccp 586 get products

### PR DESCRIPTION
**Pull request by sofib**
Working implementation for get products, contains fixes for filters and more parameters to the fact-finder search itself.

### CODE REVIEW

File: extension/lib/getProducts.search.js [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    if (input.filters) {
+      Object.keys(input.filters).forEach(filter => {
+        const shopgatefilter = input.filters[filter]
+        // atm we support only multi select
+        searchRequest.addFilter(filter, FactFinderClient.groups.filterStyle.MULTISELECT, shopgatefilter.values)
+      })
+    }
+
+    // pagination - limit and offset are optional
+    if (input.limit && input.offset) {
```
Comment (chrisEff) [inline extension/lib/getProducts.search.js]: If input is present, but offset not, how about assuming zero as a default offset?

File: extension/lib/getProducts.search.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+    const followSearch = await context.storage.device.get(getFollowSearchKey(searchHash))
+    if (followSearch) {
+      context.log.debug(decorateDebug(getCollectables(context, input, { followSearch })), 'Following search')
+      searchRequest.followSearch(followSearch)
+    }
+
+    const searchResults = await factFinderClient.search(searchRequest.build(), context.config.uidTemplate)
+
+    // store the followSearch parameter
+    context.storage.device.set(getFollowSearchKey(searchHash), searchResults.followSearch)
```
Comment (chrisEff) [inline extension/lib/getProducts.search.js]: I think we should await this and handle the error with a regular try/catch

File: extension/lib/shopgate/logDecorator.js [hunk 3]
```diff
@@ -0,0 +1,41 @@
+const EXTENSION = '@shopgate-search-fact-finder'
+
+module.exports = {
+  /**
+   * @param {Error} err
+   * @param {string} [importance='high']
+   * @returns {Object}
+   */
+  decorateError (err, importance = 'high') {
```
Comment (chrisEff) [inline extension/lib/shopgate/logDecorator.js]: importance? i think usually this is rather called "priority" or "level"...
Also, "high" sounds pretty uncommon as a log level.

File: extension/lib/getProducts.search.js (hunk 1 above)
Comment (sofib) [reply extension/lib/getProducts.search.js]: Fact-finder client as a default has page = 1, we don not need to have defaults here as well, it makes the code simpler.

File: extension/lib/getProducts.search.js (hunk 2 above)
Comment (sofib) [reply extension/lib/getProducts.search.js]: We could talk on this.

File: extension/lib/getProducts.search.js [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+    context.log.error(decorateError(err), 'Search failed')
+    throw err
+  }
+}
+
+/**
+ * @param {string} searchHash
+ * @return {string}
+ */
+function getFollowSearchKey (searchHash) {
```
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/getProducts.search.js]: It is not necessary, but to me it appeared more readable this way. What do you think?

File: extension/lib/getProducts.search.js [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+/**
+ * @param {ShopgateProductSearchSort} sort
+ * @return {{fieldName: string, direction: string|null}}
+ */
+function getFactFinderSortFieldNameAndDirection (sort) {
+  // TODO there might be mapping here necessary eg price => PREIS
+  return {
+    fieldName: 'PREIS',
```
[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/getProducts.search.js]: I suppose not - but haven't verified, therefore the todo.

File: extension/lib/shopgate/logDecorator.js (hunk 3 above)
Comment (sofib) [reply extension/lib/shopgate/logDecorator.js]: Well, yes usually it is, but this is not the log level here, log level is error and importance is there to enable us to put additional information when we'd like to group errors in different importance levels to order which ones we should solve first, therefore no standard was used. I believe you have also seen this approach in other projects we've had. We could talk more if you wish.

File: extension/lib/getProducts.search.js (hunk 1 above)
Comment (chrisEff) [reply extension/lib/getProducts.search.js]: ok

File: extension/lib/getProducts.search.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (chrisEff) [reply extension/lib/getProducts.search.js]: well personally, i don't really find it more readable. but it also doesn't bother me. so if you wanna keep it, that's ok.

[the remark this answers is not in the document]
Comment (sofib) [reply extension/lib/getProducts.search.js]: I've changed it to a local variable, it is also not as important for me.

File: extension/test/unit/getProducts.search-test.js [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+      baseUri: 'https://www.shopgate.com/FactFinder',
+      channel: 'de',
+      uidTemplate: '{$.id}'
+    },
+    storage: {
+      device: {}
+    }
+  }
+
+  let clientStub
```
Comment (chrisEff) [inline extension/test/unit/getProducts.search-test.js]: please add this here:
/** @var {FactFinderClient} */

File: extension/test/unit/getProducts.search-test.js [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+    config: {
+      baseUri: 'https://www.shopgate.com/FactFinder',
+      channel: 'de',
+      uidTemplate: '{$.id}'
+    },
+    storage: {
+      device: {}
+    }
+  }
+
```
Comment (chrisEff) [inline extension/test/unit/getProducts.search-test.js]: please add this here:
/** @var {FactFinderClient} */

File: extension/lib/shopgate/logDecorator.js (hunk 3 above)
Comment (chrisEff) [reply extension/lib/shopgate/logDecorator.js]: ok

File: extension/test/unit/getProducts.search-test.js (hunk 7 above)
Comment (sofib) [reply extension/test/unit/getProducts.search-test.js]: done

File: extension/lib/shopgate/logDecorator.js (hunk 3 above)
Comment (sofib) [reply extension/lib/shopgate/logDecorator.js]: I've actually had a second thought after we spoke, and I think in this project we might not need additional importance indicator, so I've removed those references.

**Pull request merged**